src/libmeasurement_kit/neubot/dash_impl.hpp

```diff
@@ -0,0 +1,345 @@
+ */
+template <MK_MOCK_AS(http::request_sendrecv, http_request_sendrecv)>
+void negotiate_loop_(Var<report::Entry> entry, Var<net::Transport> txp,
+                     Settings settings, Var<Reactor> reactor,
+                     Var<Logger> logger, Callback<Error, std::string> callback,
+                     int iteration = 0, std::string auth_token = "") {
+    report::Entry value = {{"dash_rates", dash_rates()}};
+    std::string body = value.dump();
+    settings["http/path"] = "/negotiate/dash";
+    settings["http/method"] = "POST";
+    if (iteration > MAX_NEGOTIATIONS) {
+        logger->warn("neubot: too many negotiations");
+        callback(TooManyNegotiationsError(), "");
+        return;
+    }
+    http_request_sendrecv(
+        txp, settings,
+        {
+            {"Content-Type", "application/json"}, {"Authorization", auth_token},
+        },
+        body,
+        [=](Error error, Var<http::Response> res) {
+            if (error) {
+                logger->warn("neubot: negotiate failed: %s",
+                             error.explain().c_str());
+                callback(error, "");
+                return;
+            }
+            assert(!!res);
+            if (res->status_code != 200) {
+                logger->warn("neubot: negotiate failed on server side");
+                callback(http::HttpRequestFailedError(), "");
+                return;
+            }
+            nlohmann::json respbody;
+            std::string auth;
+            int queue_pos = 0;
+            std::string real_address;
+            int unchoked = 0;
+            try {
+                respbody = nlohmann::json::parse(res->body);
+                auth = respbody.at("authorization");
+                queue_pos = respbody.at("queue_pos");
+                real_address = respbody.at("real_address");
+                unchoked = respbody.at("unchoked");
+            } catch (const std::exception &exc) {
+                logger->warn("neubot: cannot parse negotiate response: %s",
+                             exc.what());
+                callback(CannotParseNegotiateResponseError(), "");
+                return;
+            }
+            logger->debug("negotiation: unchoked=%d queue_pos=%d", unchoked,
+                          queue_pos);
+            if (!unchoked) {
+                reactor->call_soon([=]() {
+                    negotiate_loop_(entry, txp, settings, reactor, logger,
+                                    callback, iteration + 1, auth);
+                });
+                return;
+            }
+            callback(NoError(), auth);
+        },
+        reactor, logger);
+}
+
+template <MK_MOCK_AS(http::request_sendrecv, http_request_sendrecv)>
+void collect_(Var<net::Transport> txp, Var<report::Entry> entry,
+              std::string auth, Settings settings, Var<Reactor> reactor,
+              Var<Logger> logger, Callback<Error> cb) {
+    std::string body = entry->dump();
```

Comment:
Problem: the original format of the entry was exactly the one of Neubot. Then I have basically modified it such that, now, what we send is not what Neubot would except. I guess here what we should dump is the specific key that contains the list of results rather than the whole object.